src/cargo/ops/cargo_install.rs

```diff
@@ -384,7 +384,8 @@ fn select_pkg<'a, T>(mut source: T,
                 None => None,
             };
             let vers = vers.as_ref().map(|s| &**s);
-            let dep = Dependency::parse_no_deprecated(name, vers, source.source_id())?;
+            let dep = Dependency::parse_no_deprecated(name, vers, source.source_id(),
+                                                      source.source_id())?;
```

Comment:
Here and a couple of other places I'm just setting the registry ID to the source ID since it's easier and we won't end up in the publish codepath from here. It's a bit gross though. Maybe it should be optional?